middle_end/flambda2/simplify/unboxing/unboxing_epa.ml

```diff
@@ -342,21 +361,49 @@ and compute_extra_args_for_variant ~pass rewrite_id ~typing_env_at_use
   in
   let fields_by_tag =
     Tag.Scannable.Map.mapi
-      (fun tag_decision block_fields ->
+      (fun tag_decision (shape, block_fields) ->
         let size = List.length block_fields in
         (* See doc/unboxing.md about invalid constants, poison and aliases. *)
         let poison_const = Const.const_int (Targetint_31_63.of_int 0xbaba) in
-        let bak : Flambda_primitive.Block_access_kind.t =
-          Values
-            { size = Known (Targetint_31_63.of_int size);
-              tag = Known tag_decision;
-              field_kind = Any_value
-            }
+        let bak index : Flambda_primitive.Block_access_kind.t =
+          match (shape : K.Block_shape.t) with
+          | Value_only ->
+            Values
+              { size = Known (Targetint_31_63.of_int size);
+                tag = Known tag_decision;
+                field_kind = Any_value
+              }
+          | Float_record ->
+            (* CR vlaviron: I suspect that this case is unreachable. At least
+               the previous version of this code didn't handle it, and it seems
+               likely that float records were handled by the unique tag and size
+               case instead. *)
+            Naked_floats { size = Known (Targetint_31_63.of_int size) }
+          | Mixed_record shape ->
+            let field_kind =
+              let field_kind =
+                (K.Mixed_block_shape.field_kinds shape).(index)
+              in
+              if index < K.Mixed_block_shape.value_prefix_size shape
+              then
+                (* CR vlaviron: we're not trying to infer if this can only be an
+                   immediate. In most cases it should be fine, as the primitive
+                   will get simplified away. *)
+                P.Mixed_block_access_field_kind.Value_prefix Any_value
+              else P.Mixed_block_access_field_kind.Flat_suffix field_kind
+            in
```

Comment:
This code is duplicated with earlier code. maybe make a `block_access_kind_of_field` function